keps/sig-cli/3104-introduce-kuberc/README.md

```diff
@@ -0,0 +1,849 @@
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Introduce a `kuberc` file as a place for client preferences.
+* Version and structure this file in a way that makes it easy to introduce new behaviors and settings for users.
+* Enable users to define kubectl command aliases.
+* Enable users to define kubectl default flags.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose introducing a new file that will be versioned and intended for user provided preferences. This new file is entirely opt-in.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+| Risk | Impact | Mitigation |
+| --- | --- | --- |
+| Confusing users with a new config file | Low | Documentation and education |
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The file will default to being located in `~/.kube/kuberc`. A flag will allow overriding this default location with a path i.e. `kubectl --kuberc /var/kube/rc`.
```

Comment:
is this an opportunity to address some of the issues raised in https://github.com/kubernetes/kubernetes/issues/80399, https://github.com/kubernetes/enhancements/issues/2229, or https://github.com/kubernetes/enhancements/pull/2111 for this new file?

Reply:
We opted out of XDG because it would cause an awful partition across tooling that didn't upgrade client-go.

@BenTheElder had an idea on how we can introduce a `~/.kube/config.d` folder separate from this.

Reply:
> We opted out of XDG because it would cause an awful partition across tooling that didn't upgrade client-go.

For existing files, sure. For this net new file, it's worth considering, at least.

Reply:
@soltysh what are your thoughts here? My first thought is introducing a new location will just confuse users when kubeconfigs will not be loaded from there.

Reply:
I think it might be worth explicitly addressing in the KEP but I also lean towards it being confusing to have two sets of rules as to what directory to use as for reasoning why we aren't doing that.

Reply:
I agree with Ben, referring to that XDG (especially it's there as rejected) and make it explicit why we choose this approach is good idea.